### PR TITLE
WIP: Fix compiler warnings

### PR DESCRIPTION
**Pull request by stv0g**
This PR fixes a lot of compiler warnings thrown by newer GCC (>= 6) versions…

This are mostly warnings about:
 - comparison differing in signedness
 - unused functions and variables
 - some casts

I found these by building the spectrum Docker image using Fedora 27 as baseimage.
Another PR for my improvements of the Docker image will follow.

### CODE REVIEW

**Comment by vitalyster** [comment]
`std::unique_ptr` isn't available in c++03 mode, which is required to be compatible with swiften 2 and 3 releases (c++11 swiften 4 isn't released yet)

**Comment by vitalyster** [comment]
conditional defines based on swiften version are here - https://github.com/SpectrumIM/spectrum2/blob/master/include/Swiften/SwiftenCompat.h

File: backends/libpurple/main.cpp [hunk 1]
```diff
@@ -1440,6 +1440,7 @@ struct Dis {
 	std::string protocol;
 };
 
+__attribute__ ((unused))
```
Comment (vitalyster) [inline backends/libpurple/main.cpp]: Looks like GCC-specific and will not work with Clang/MSVC

Comment (stv0g) [reply backends/libpurple/main.cpp]: This attribute is also [supported by Clang](https://clang.llvm.org/docs/AttributeReference.html#maybe-unused-unused-gnu-unused):

```
int main(int argc, char *argv[]) {
	return 0;
}

__attribute__ ((unused)) static int bla() {
	return 42;
}
```

```
$ clang -Wall test.cpp
```

MSVC seems to have problems with the attribute declaration.
But there is a simple workaround: https://stackoverflow.com/questions/19249521/attribute-macro-and-its-effect-on-visual-studio-2010-based-projects

**Comment by stv0g** [comment]
I found a [couple more warnings in the Clang / OS X run via Travis-CI](https://travis-ci.org/SpectrumIM/spectrum2/jobs/294881921).

Let me fix those as well before merging the PR

**Comment by vitalyster** [comment]
I have fixed remaining errors, thanks

**Comment by stv0g** [comment]
Thanks :)

**Pull request merged**